dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1196 @@
+	peersTried := make(map[peer.ID]struct{})
+
+	// test base case
+	cs.peerState[testPeerA] = &peerState{
+		number: 129,
+	}
+	cs.peerState[testPeerB] = &peerState{
+		number: 257,
+	}
+
+	peers := cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 2, len(peers))
+	require.Contains(t, peers, testPeerA)
+	require.Contains(t, peers, testPeerB)
+
+	// test peer ignored case
+	cs.ignorePeers[testPeerA] = struct{}{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+
+	// test all peers ignored case
+	cs.ignorePeers[testPeerB] = struct{}{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 2, len(peers))
+	require.Contains(t, peers, testPeerA)
+	require.Contains(t, peers, testPeerB)
+	require.Equal(t, 0, len(cs.ignorePeers))
+
+	// test peer's best block below number case, shouldn't include that peer
+	start, err := variadic.NewUint32OrHash(130)
+	require.NoError(t, err)
+	req.StartingBlock = *start
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+
+	// test peer tried case, should ignore peer already tried
+	peersTried[testPeerA] = struct{}{}
+	req.StartingBlock = variadic.Uint32OrHash{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+}
+
+func Test_chainSync_logSyncSpeed(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+
+	mockBlockState := NewMockBlockState(ctrl)
+	mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+	mockBlockState.EXPECT().GetHighestFinalisedHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+
+	mockNetwork := NewMockNetwork(ctrl)
+	mockNetwork.EXPECT().Peers().Return([]common.PeerInfo{}).AnyTimes()
+
+	type fields struct {
+		blockState  BlockState
+		network     Network
+		state       chainSyncState
+		benchmarker *syncBenchmarker
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "state bootstrap",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       bootstrap,
+			},
+		},
+		{
+			name: "case tip",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       tip,
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithCancel(context.Background())
+			cs := &chainSync{
+				ctx:           ctx,
+				cancel:        cancel,
+				blockState:    tt.fields.blockState,
+				network:       tt.fields.network,
+				state:         tt.fields.state,
+				benchmarker:   tt.fields.benchmarker,
+				logSyncPeriod: time.Nanosecond,
+			}
+			go cs.logSyncSpeed()
+			time.Sleep(time.Nanosecond)
```

Comment:
We need this event driven, that will likely fail occasionally.

If the prod code can't be changed to be event driven (i.e. with signal channels), use the last mock call together with a `DoAndReturn` and a signal channel to know when we can cancel and finish operations.

Reply:
Added wait group to test.